### PR TITLE
Bypass MPolynomial Representation of AIR Constraints

### LINKED ISSUE #96: Bypass MPolynomial Representation of AIR Constraints
Except for the Hash Table, all constraints are represented as multivariate polynomials using the `MPolynomial` struct. This representation is both more reflective of the underlying maths that makes STARKs work and more flexible for a system whose construction is in flux. However, this representation is also one of the main reasons why the prover and verifier are slow. By bypassing the multivariate polynomial representation for evaluating the constraints we can gain significant speedups.

Note that the multivariate polynomial representations are still useful for testing and debugging. In particular, you can evaluate the hardcoded evaluator and the symbolic polynomials in a random point; if the two values coincide then with high confidence both express the same constraints.

---

- [x] initial constraints
- [x] consistency constraints
- [x] transition constraints
- [x] terminal constraints

### PR DESCRIPTION
**Pull request by jan-ferdinand**
Summary:
- express all constraints of all tables as arithmetic circuits
- generate code to evaluate those constraints
- generate code for getting quotient degree bounds
- remove memoization of AIR constraints & degree bounds
- $\Rightarrow$ completely remove multivariate polynomials from tables

Running `cargo criterion --bench prove_fib_100` on this PR's branch comparing to current `master`:

Fix #96, #121, #123

### CODE REVIEW

**Review by aszepieniec — CHANGES_REQUESTED** [review]
Looks good to me.

Requested changes:

 - build script that automatically generates constraints code when not already present
 - ~~don't spam terminal when generating code~~
 - test suite should work out of the box

File: triton-vm/src/table/extension_table.rs [hunk 1]
```diff
@@ -23,97 +23,53 @@ use super::challenges::AllChallenges;
...
+const ERROR_MESSAGE_GENERATE_CONSTRAINTS: &str =
+    "Constraints must be in place. Run: `cargo run --bin constraint-evaluation-generator`";
+const ERROR_MESSAGE_GENERATE_DEGREE_BOUNDS: &str =
+    "Degree bounds must be in place. Run: `cargo run --bin constraint-evaluation-generator`";
 
 pub trait Evaluable: ExtensionTable {
-    /// Evaluate initial constraints if they are set. This default impl
-    /// can be overridden by running the `constraint-evaluation-generator`.
+    /// The code for this method must be generated by running
+    /// `OUTPUT_RUST_SOURCE_CODE=1 cargo run --bin constraint-evaluation-generator`
```
Comment (sshine) [inline triton-vm/src/table/extension_table.rs:35]:
```suggestion
    /// `cargo run --bin constraint-evaluation-generator`
```

File: triton-vm/src/table/extension_table.rs [hunk 2]
```diff
@@ -23,97 +23,53 @@ use super::challenges::AllChallenges;
...
-        } else {
-            panic!("{} does not have initial constraints!", &self.name());
-        }
+        panic!("{ERROR_MESSAGE_GENERATE_CONSTRAINTS}")
     }
 
-    /// Evaluate consistency constraints if they are set. This default impl
-    /// can be overridden by running the `constraint-evaluation-generator`.
+    /// The code for this method must be generated by running
+    /// `OUTPUT_RUST_SOURCE_CODE=1 cargo run --bin constraint-evaluation-generator`
```
Comment (sshine) [inline triton-vm/src/table/extension_table.rs:45]:
```suggestion
    /// `cargo run --bin constraint-evaluation-generator`
```

File: triton-vm/src/table/extension_table.rs [hunk 3]
```diff
@@ -23,97 +23,53 @@ use super::challenges::AllChallenges;
...
-        } else {
-            panic!("{} does not have consistency constraints!", &self.name());
-        }
+        panic!("{ERROR_MESSAGE_GENERATE_CONSTRAINTS}")
     }
 
-    /// Evaluate transition constraints if they are set. This default impl
-    /// can be overridden by running the `constraint-evaluation-generator`.
+    /// The code for this method must be generated by running
+    /// `OUTPUT_RUST_SOURCE_CODE=1 cargo run --bin constraint-evaluation-generator`
```
Comment (sshine) [inline triton-vm/src/table/extension_table.rs:55]:
```suggestion
    /// `cargo run --bin constraint-evaluation-generator`
```

File: triton-vm/src/table/extension_table.rs [hunk 4]
```diff
@@ -23,97 +23,53 @@ use super::challenges::AllChallenges;
...
-        } else {
-            panic!("{} does not have transition constraints!", &self.name());
-        }
+        panic!("{ERROR_MESSAGE_GENERATE_CONSTRAINTS}")
     }
 
-    /// Evaluate terminal constraints if they are set. This default impl
-    /// can be overridden by running the `constraint-evaluation-generator`.
+    /// The code for this method must be generated by running
+    /// `OUTPUT_RUST_SOURCE_CODE=1 cargo run --bin constraint-evaluation-generator`
```
Comment (sshine) [inline triton-vm/src/table/extension_table.rs:66]:
```suggestion
    /// `cargo run --bin constraint-evaluation-generator`
```

**Review by sshine — APPROVED** [review]

**Pull request merged**